packages/admin-server/src/app.ts

```diff
@@ -177,6 +189,108 @@ export const createApp = async (): Promise<{app: Koa, config: Config}> => {
     rewrite: path => path.replace(staticPath, '/')
   }));
 
+  const sessionTokenCacheExpireTime = 30;
+  const sessionTokenCacheStore = new NodeCache({stdTTL: sessionTokenCacheExpireTime});
+  const checkSessionToken = async (ctx: Koa.ParameterizedContext) => {
+    const sessionToken = ctx.cookies.get('phapplication-session-id') || '';
+    const sessionTokenCached = sessionTokenCacheStore.get(sessionToken);
+    const scope = ctx.params.scope || '';
+    if (sessionToken) {
+      if (sessionTokenCached) {
+        return true;
+      }
+      ctx.resetSessionToken = true;
+    }
+    return false
+  };
+
+  const updateSessionToken = async (ctx: Koa.ParameterizedContext) => {
+    const appID = ctx.params.appID;
+    let sessionToken = ctx.cookies.get('phapplication-session-id');
+    if (!sessionToken || ctx.resetSessionToken) {
+      // Generate session token
+      sessionToken = uuidv4();
+      console.log('Generate token');
+      console.log(sessionToken);
+      ctx.cookies.set('phapplication-session-id', sessionToken, {path: `${config.appPrefix}/apps/${appID}`});
+    }
+    console.log('Update Token: ', sessionToken);
+    sessionTokenCacheStore.set(sessionToken, true, sessionTokenCacheExpireTime);
+
+    return true;
+  };
+
+  const checkUserGroup = async (ctx: Koa.ParameterizedContext) => {
+    const {userId} = oidcCtrl.getUserFromContext(ctx);
+    const variables = {
+      id: userId,
+    };
+    const query = gql`
+    query ($id: ID!) {
+      user(where: {id: $id}) {
+        id
+        username
+        groups {name}
+      }
+    }
+    `;
+    const data = await ctx.graphqlClient.request(query, variables);
+
+    for (const group of data.user.groups) {
+      if (group.name === 'phusers') {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  const phApplicationProxyHandler = async (ctx: Koa.ParameterizedContext, next: any) => {
+    const appID = ctx.params.appID;
+    const scope = ctx.params.appID;
+    // const service = 'app-mlflow-xyzab.hub.svc.cluster.local:5000';
+    const service = 'localhost:5000';
+    ctx.params.scope = scope;
+    ctx.params.service = service;
+
+    console.log(`${config.appPrefix}/apps/${appID}`);
+    switch (scope) {
+      case 'group':
+      case 'primehub':
+        const sessionCached = await checkSessionToken(ctx);
+        if (sessionCached === false) {
+          console.log('Token: ', ctx.cookies.get('phapplication-session-id'));
+          return oidcCtrl.loggedIn(ctx, async () => {
+            if (scope === 'group') {
+              const groupAllowed = await checkUserGroup(ctx);
+              if (groupAllowed === false) {
+                throw Boom.forbidden('Request not authorized');
+              }
+            }
+            updateSessionToken(ctx);
```

Comment:
It seems that we don't reset the session timeout for each request? we can use https://github.com/node-cache/node-cache#change-ttl-ttl to update the ttl

Reply:
Fixed by adding new function call `updateSessionTTL` when `sessionToken` cache hit.